docs/When_Do_You_Need_A_Custom_Device.md

```diff
@@ -0,0 +1,73 @@
+ 
+Can the hardware requirement be met by passing LabVIEW DBLs to and from the custom device during steady state operation?
+If the hardware driver returns a vector, structure, or any non-DBL data, it cannot be passed directly from the custom device to the rest of the VeriStand system. The developer is responsible for coercing the data (or using an alternative communication mechanism) to pass data from the custom device to the rest of the system. For more information on the available communication mechanisms, see **[LabVIEW 2018 Real-Time Module Help](https://zone.ni.com/reference/en-XX/help/370715P-01/)** » **[Real-Time Module Concepts](https://zone.ni.com/reference/en-XX/help/370715P-01/lvrtconcepts/rtmoduleconcepts/)** » **[Sharing Data in Deterministic Applications](https://zone.ni.com/reference/en-XX/help/370715P-01/lvrtconcepts/rtsharedatadeterm/)** » **[Exploring Remote Communication Methods](https://zone.ni.com/reference/en-XX/help/370715P-01/lvrtconcepts/exploring_communication_methods/)**.
+VeriStand also exposes its TCP pipe via dynamic event registration. This pipe may suite your remote communication requirements. See the Custom Device Engine Events section for more information.
+
+#### Testing
+
+A custom device is one part of a VeriStand system. The complete state of the operator's system is seldom known by the custom device developer. System state includes the following information.
+1 What are the specifications of the Target and Host Computer? Check the **[Glossary](https://www.ni.com/documentation/en/veristand/latest/manual/veristand-glossary/)** section of the **[VeriStand Help](https://www.ni.com/documentation/en/veristand/latest/manual/manual-overview/)**.
+2 What components are in the system definition?
+3 How computationally intense are the simulation models?
+4 What loop rates are required?
+5 What is the health and resource utilization of the system?
+Ideally, the custom device is implemented to be minimally burdensome, extremely efficient, and easy to use. Depending on its complexity, it may become necessary to test, debug, and optimize the code on systems representative of the operator’s system. Consider the following example.
+A custom device developer needs to benchmark a 3rd-party hardware custom device. He adds the custom device to the Engine Demo example that ships with VeriStand. He deploys the system definition to a quad-core PXIe-8880 Controller. Adding the custom device to the system increased the target’s CPU load by 10% per-core and RAM utilization increased 120KB. If the operator is deploying the same custom device to a single-core PXIe-8821 Controller, with an average CPU load of 60% because of a computationally intense model, it’s unlikely the operator will achieve the same loop rate after adding the custom device. This system may be incapable of running the custom device at all.
```

Comment:
```suggestion
Consider the following example. A custom device developer needs to benchmark a 3rd-party hardware custom device. He adds the custom device to the Engine Demo example that ships with VeriStand. He deploys the system definition to a quad-core PXIe-8880 Controller. Adding the custom device to the system increased the target’s CPU load by 10% per-core and RAM utilization increased 120KB. If the operator is deploying the same custom device to a single-core PXIe-8821 Controller, with an average CPU load of 60% because of a computationally intense model, it’s unlikely the operator will achieve the same loop rate after adding the custom device. This system may be incapable of running the custom device at all.
```